streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -261,12 +260,18 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
                 taskManager.suspendedActiveTaskIds(),
                 taskManager.suspendedStandbyTaskIds());
 
+            if (streamThread.assignmentErrorCode.get() == StreamsPartitionAssignor.Error.INCOMPLETE_SOURCE_TOPIC_METADATA.getCode()) {
+                log.debug("Received error code " + streamThread.assignmentErrorCode.get() +
+                    " - shutdown");
+                streamThread.shutdown();
+                streamThread.setStateListener(null);
```

Comment:
@guozhangwang What is the reasoning behind this?

Reply:
Copying my response from the email thread:

```
I wanted the integration test in the same PR since I'm concerned about the call trace of `streamThread.shutdown()` in the onPartitionAssigned callback, because today we only call that in the main loop of the thread, I'm concerned if there are any race condition that may cause deadlocks.. The integration test should not actually be hard to add, you can take a look at the integration test folder and let me know if you have any questions.
```

Currently I cannot think of a race condition that calling `shutdown` in the callback would introduce than calling shutdown in the main loop, but I'm not 100% sure, so I insisted on triggering a system test.